### ISSUE TITLE
detect invalid nodes and invalid graph in onnx checker

### ISSUE THREAD

**Issue by jeyblu** [onnx/onnx#3629]
# Bug Report

### Describe the bug
onnxruntime uses the onnx [checker](https://github.com/onnx/onnx/blob/master/onnx/checker.cc) to validate graph. Checker didn't catch a test case, [TEST(BatchNormTest, InvalidScaleDim)](https://github.com/microsoft/onnxruntime/blob/378a98597e2c61ccd11343906bd8d83d1af6b7c7/onnxruntime/test/providers/cpu/nn/batch_norm_op_test.cc#L406), where the node is invalid.

### System information
- OS Platform and Distribution (*e.g. Linux Ubuntu 16.04*): Windows 10
- ONNX version (*e.g. 1.7*):  master
- Python version: 3.7.7
- GCC/Compiler version (if compiling from source):
- CMake version: 3.20.3
- Protobuf version:
- Visual Studio version (if applicable): 2019

### Reproduction instructions
- Build onnxruntime for oneDNN execution provider
- Run onnxruntime_test_all -e dnnl --gtest_filter=BatchNormTest.InvalidScaleDim

### Expected behavior
onnx checker detects that the node in the test is invalid

### Notes
Related discussion in onnxruntime: [8152](https://github.com/microsoft/onnxruntime/discussions/8152)

**Comment by askhade** [comment]
@jeyblu : Please fell free to propose a fix to checker.

**Comment by gramalingam** [comment]
Agree that such a check can/should be added to the shape-inference method for the op.

**Comment by jeyblu** [comment]
We pulled the latest onnx code that contains [Batchnorm shape inference (#3657)](https://github.com/onnx/onnx/commit/c24cd429c4ab47d2b057da8174788c39c60f8760) into the onnx submodule in onnxruntime. Ran the onnxruntime unit test [TEST(BatchNormTest, InvalidScaleDim)](https://github.com/microsoft/onnxruntime/blob/378a98597e2c61ccd11343906bd8d83d1af6b7c7/onnxruntime/test/providers/cpu/nn/batch_norm_op_test.cc#L406). The onnx checker is still not catching the invalid scale dimension.

onnxruntime  [graph.cc](https://github.com/microsoft/onnxruntime/blob/d14b08d09cee43b42fa9ac361a78d1cbdbceddef/onnxruntime/core/graph/graph.cc#L2342) calls the onnx [check_node](https://github.com/onnx/onnx/blob/c24cd429c4ab47d2b057da8174788c39c60f8760/onnx/checker.cc#L529) function in checker.cpp. The call stack shows that [check_node](https://github.com/onnx/onnx/blob/c24cd429c4ab47d2b057da8174788c39c60f8760/onnx/checker.cc#L551) calls [check_attribute](https://github.com/onnx/onnx/blob/c24cd429c4ab47d2b057da8174788c39c60f8760/onnx/checker.cc#L439). The [comment](https://github.com/onnx/onnx/blob/c24cd429c4ab47d2b057da8174788c39c60f8760/onnx/checker.cc#L437) for check_attribute says
```
// NB: This is a generic "attribute well-formedness" check, it doesn't
// actually test if an attribute is valid per a schema
```

stack
```>	onnxruntime_test_all.exe!onnx::checker::check_attribute(const onnx::AttributeProto & attr, const onnx::checker::CheckerContext & ctx, const onnx::checker::LexicalScopeContext & lex_ctx) Line 439	C++
 	onnxruntime_test_all.exe!onnx::checker::check_node(const onnx::NodeProto & node, const onnx::checker::CheckerContext & ctx, const onnx::checker::LexicalScopeContext & lex_ctx) Line 552	C++
 	onnxruntime_test_all.exe!onnxruntime::Graph::VerifyNodeAndOpMatch(const onnxruntime::Graph::ResolveOptions & options) Line 2350	C++
 	onnxruntime_test_all.exe!onnxruntime::Graph::PerformTypeAndShapeInferencing(const onnxruntime::Graph::ResolveOptions & options) Line 2524	C++
 	onnxruntime_test_all.exe!onnxruntime::Graph::Resolve(const onnxruntime::Graph::ResolveOptions & options) Line 2591	C++
 	onnxruntime_test_all.exe!onnxruntime::test::OpTester::Run(onnxruntime::SessionOptions so, onnxruntime::test::OpTester::ExpectResult expect_result, const std::string & expected_failure_string, const std::unordered_set<std::string,std::hash<std::string>,std::equal_to<std::string>,std::allocator<std::string>> & excluded_provider_types, const OrtRunOptions * run_options, std::vector<std::unique_ptr<onnxruntime::IExecutionProvider,std::default_delete<onnxruntime::IExecutionProvider>>,std::allocator<std::unique_ptr<onnxruntime::IExecutionProvider,std::default_delete<onnxruntime::IExecutionProvider>>>> * execution_providers, const onnxruntime::Graph::ResolveOptions & options, unsigned __int64 * number_of_pre_packed_weights_counter, unsigned __int64 * number_of_shared_pre_packed_weights_counter) Line 943	C++
 	onnxruntime_test_all.exe!onnxruntime::test::OpTester::Run(onnxruntime::test::OpTester::ExpectResult expect_result, const std::string & expected_failure_string, const std::unordered_set<std::string,std::hash<std::string>,std::equal_to<std::string>,std::allocator<std::string>> & excluded_provider_types, const OrtRunOptions * run_options, std::vector<std::unique_ptr<onnxruntime::IExecutionProvider,std::default_delete<onnxruntime::IExecutionProvider>>,std::allocator<std::unique_ptr<onnxruntime::IExecutionProvider,std::default_delete<onnxruntime::IExecutionProvider>>>> * execution_providers, ExecutionMode execution_mode, const onnxruntime::Graph::ResolveOptions & options) Line 891	C++
 	onnxruntime_test_all.exe!onnxruntime::test::TestBatchNorm<float>(const std::unordered_map<std::string,std::vector<float,std::allocator<float>>,std::hash<std::string>,std::equal_to<std::string>,std::allocator<std::pair<std::string const ,std::vector<float,std::allocator<float>>>>> & input_data_map, const std::unordered_map<std::string,std::vector<__int64,std::allocator<__int64>>,std::hash<std::string>,std::equal_to<std::string>,std::allocator<std::pair<std::string const ,std::vector<__int64,std::allocator<__int64>>>>> & input_shapes_map, std::optional<float> epsilon, const std::initializer_list<float> & expected_output, const std::vector<__int64,std::allocator<__int64>> & expected_output_shape, bool all_input_except_x_are_initializers, __int64 spatial_mode, onnxruntime::test::OpTester::ExpectResult expect_result, const std::string & err_str, int opset_version) Line 53	C++
 	onnxruntime_test_all.exe!onnxruntime::test::BatchNormTest_InvalidScaleDim_Test::TestBody() Line 439	C++
 	onnxruntime_test_all.exe!testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test,void>(testing::Test * object, void(testing::Test::*)() method, const char * location) Line 2592	C++
 	onnxruntime_test_all.exe!testing::internal::HandleExceptionsInMethodIfSupported<testing::Test,void>(testing::Test * object, void(testing::Test::*)() method, const char * location) Line 2643	C++
 	onnxruntime_test_all.exe!testing::Test::Run() Line 2689	C++
 	onnxruntime_test_all.exe!testing::TestInfo::Run() Line 2864	C++
 	onnxruntime_test_all.exe!testing::TestSuite::Run() Line 3016	C++
 	onnxruntime_test_all.exe!testing::internal::UnitTestImpl::RunAllTests() Line 5856	C++
 	onnxruntime_test_all.exe!testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl,bool>(testing::internal::UnitTestImpl * object, bool(testing::internal::UnitTestImpl::*)() method, const char * location) Line 2592	C++
 	onnxruntime_test_all.exe!testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl,bool>(testing::internal::UnitTestImpl * object, bool(testing::internal::UnitTestImpl::*)() method, const char * location) Line 2643	C++
 	onnxruntime_test_all.exe!testing::UnitTest::Run() Line 5438	C++
 	onnxruntime_test_all.exe!RUN_ALL_TESTS() Line 2491	C++
 	onnxruntime_test_all.exe!main(int argc, char * * argv) Line 45	C++
 	[External Code]	
```

**Comment by gramalingam** [comment]
This check should happen when the type/shape inference method is invoked for the node (not when the generic check is invoked). I thought that Graph::Resolve does call the type/shape inference method, so I am puzzled why that wouldn't get called.

**Comment by jeyblu** [comment]
[Resolve](https://github.com/microsoft/onnxruntime/blob/0983d619692a39a2135eddf5faa6d80a6c6bcb80/onnxruntime/core/graph/graph.cc#L2550) calls [PerformTypeAndShapeInferencing](https://github.com/microsoft/onnxruntime/blob/0983d619692a39a2135eddf5faa6d80a6c6bcb80/onnxruntime/core/graph/graph.cc#L2509), which calls [TypeCheckInputsAndInitializers](https://github.com/microsoft/onnxruntime/blob/0983d619692a39a2135eddf5faa6d80a6c6bcb80/onnxruntime/core/graph/graph.cc#L2247). Is the code that calls the function to verify the scale dimension in [TypeCheckInputsAndInitializers](https://github.com/microsoft/onnxruntime/blob/0983d619692a39a2135eddf5faa6d80a6c6bcb80/onnxruntime/core/graph/graph.cc#L2247)?

[PerformTypeAndShapeInferencing](https://github.com/microsoft/onnxruntime/blob/0983d619692a39a2135eddf5faa6d80a6c6bcb80/onnxruntime/core/graph/graph.cc#L2509) also calls [VerifyNodeAndOpMatch](https://github.com/microsoft/onnxruntime/blob/0983d619692a39a2135eddf5faa6d80a6c6bcb80/onnxruntime/core/graph/graph.cc#L2313) which calls onnx [check_node](https://github.com/onnx/onnx/blob/c24cd429c4ab47d2b057da8174788c39c60f8760/onnx/checker.cc#L529). Does calling onnx check_node not call the check the scale dimension? If not, which function in onnx [checker.cc](https://github.com/onnx/onnx/blob/master/onnx/checker.cc) will check the scale dimension?
